File: src/web/src/utils/attributeAnalysis.js
export const emptyAttributeSchema = () => ({ objectType: 'person', attributes: [] })
export const parseAttributeSchema = (value) => {
  try {
    const schema = typeof value === 'string' ? JSON.parse(value) : JSON.parse(JSON.stringify(value || emptyAttributeSchema()))
    return schema && Array.isArray(schema.attributes) ? schema : emptyAttributeSchema()
  }
  catch { return emptyAttributeSchema() }
}
const identifier = (value) => typeof value === 'string' && /^[A-Za-z0-9_.-]{1,64}$/.test(value)
export function validAttributeSchema(schema) {
  if (!identifier(schema?.objectType) || !Array.isArray(schema.attributes) || !schema.attributes.length || schema.attributes.length > 32) return false
  if (new TextEncoder().encode(JSON.stringify(schema)).length > 65536) return false
  const keys = new Set()
  return schema.attributes.every(a => {
    if (!identifier(a.key) || keys.has(a.key) || !a.name || a.name.length > 128 || !identifier(a.sourceNode) || !identifier(a.modelCode)) return false
    keys.add(a.key)
    if (!['single', 'multiple', 'binary'].includes(a.type) || !Number.isFinite(a.threshold) || a.threshold < 0 || a.threshold > 1 || !Number.isFinite(a.minRatio) || a.minRatio <= 0.5 || a.minRatio > 1) return false
    if (!Array.isArray(a.options) || !a.options.length || a.options.length > 64) return false
    const binary = a.type === 'binary'
    if (binary && (a.options.length !== 2 || !a.options[0].label || a.options[1].label !== '')) return false
    const labels = new Set(), values = new Set()
    return a.options.every(o => {
      if (typeof o.label !== 'string' || (!binary && !o.label) || o.label.length > 128 || !identifier(o.value) || !o.name || o.name.length > 128 || labels.has(o.label) || values.has(o.value)) return false
      labels.add(o.label); values.add(o.value)
      return true
    })
  })
}
export const attributeParam = (node, key) => node?.configObject?.params?.find(p => p.key === key)?.value
export function attributeNodes(process) {
  try {
    const nodes = typeof process === 'string' ? JSON.parse(process) : process
    return (nodes || []).filter(n => n.actionId === 'BA_20003' && attributeParam(n, 'inputMode') === 'attributes')
  } catch { return [] }
}
export function validAttributeFlow(process) {
  try {
    const nodes = typeof process === 'string' ? JSON.parse(process) : process
    const accumulators = attributeNodes(nodes)
    if (!accumulators.length) return true
    if (accumulators.length !== 1) return false
    const node = accumulators[0], schema = parseAttributeSchema(attributeParam(node, 'attributeSchema'))
    if (!validAttributeSchema(schema)) return false
    const byId = new Map(nodes.map(n => [String(n.flowActionId), n]))
    const ancestors = (current, seen = new Set()) => {
      for (const id of String(current?.preFlowActionId || '').split(',')) {
        if (!id || seen.has(id) || !byId.has(id)) continue
        seen.add(id); ancestors(byId.get(id), seen)
      }
      return seen
    }
    const upstream = ancestors(node)
    if ([node, ...[...upstream].map(id => byId.get(id))].some(n => String(n.preFlowActionId || '').includes(','))) return false
    return [...upstream].some(id => byId.get(id).actionId === 'AA_00003') &&
      schema.attributes.every(a => upstream.has(a.sourceNode) && byId.get(a.sourceNode)?.actionId === 'AA_00002' && String(attributeParam(byId.get(a.sourceNode), 'atomicCode')) === a.modelCode && [...ancestors(byId.get(a.sourceNode))].some(id => byId.get(id).actionId === 'AA_00003')) &&
      nodes.some(n => n.actionId === 'BA_00004' && ancestors(n).has(String(node.flowActionId)))
  } catch { return false }
}
export function attributeRecord(row) {
  try { return (typeof row.property === 'string' ? JSON.parse(row.property) : row.property)?.attributes || null }
  catch { return null }
}
export function attributeDisplay(record, definition, unknown, failed, notApplicable) {
  const value = record?.attributes?.find(a => a.key === definition.key)
  if (!value) return notApplicable
  if (value.status !== 'valid') return value.status === 'failed' ? failed : unknown
  const original = record.schema.attributes.find(a => a.key === definition.key)
  return value.values.map(v => original?.options.find(o => o.value === v)?.name || v).join(' / ')
}

export function attributeClassifierSources(nodes, edges, targetId) {
  const upstream = new Set()
  const visit = id => edges.filter(edge => String(edge.target) === String(id)).forEach(edge => {
    const parent = String(edge.source)
    if (!upstream.has(parent)) { upstream.add(parent); visit(parent) }
  })
  visit(targetId)
  return nodes.filter(node => upstream.has(String(node.id)) && node.data?.actionId === 'AA_00002')
    .map(node => ({ ...node.data?.configObject?.webConfig?.atomic, position: String(node.id) }))
    .filter(source => source.atomicCode)
}

File: src/util/AttributeAnalysis.cc
#include "util/AttributeAnalysis.h"

#include <algorithm>
#include <cmath>
#include <functional>
#include <iomanip>
#include <map>
#include <set>
#include <sstream>

namespace cosmo {
namespace {
    bool Identifier(const std::string& value) {
        if (value.empty() || value.size() > 64)
            return false;
        for (unsigned char c : value) {
            if (!((c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9') || c == '_' ||
                  c == '-' || c == '.'))
                return false;
        }
        return true;
    }
}  // namespace

bool ValidateAttributeSchema(const AttributeSchema& schema) {
    if (!Identifier(schema.objectType) || schema.attributes.empty() || schema.attributes.size() > 32)
        return false;
    std::set<std::string> keys;
    for (const auto& attr : schema.attributes) {
        if (!Identifier(attr.key) || !keys.insert(attr.key).second || attr.name.empty() ||
            attr.name.size() > 128 || !Identifier(attr.sourceNode) || !Identifier(attr.modelCode) ||
            (attr.type != "single" && attr.type != "multiple" && attr.type != "binary") ||
            !std::isfinite(attr.threshold) || attr.threshold < 0 || attr.threshold > 1 ||
            !std::isfinite(attr.minRatio) || attr.minRatio <= 0.5 || attr.minRatio > 1 ||
            attr.options.empty() || attr.options.size() > 64)
            return false;
        const bool binary = attr.type == "binary";
        if (binary && (attr.options.size() != 2 || attr.options.front().label.empty() ||
                       !attr.options.back().label.empty()))
            return false;
        std::set<std::string> labels, values;
        for (const auto& option : attr.options) {
            if ((!binary && option.label.empty()) || option.label.size() > 128 || !Identifier(option.value) ||
                option.name.empty() || option.name.size() > 128 || !labels.insert(option.label).second ||
                !values.insert(option.value).second)
                return false;
        }
    }
    return nlohmann::json(schema).dump().size() <= 65536;
}

std::string AttributeSchemaId(const AttributeSchema& schema) {
    return AttributeFingerprint(nlohmann::json(schema).dump());
}

bool ValidateAttributeRecord(const AttributeRecord& record) {
    if (!ValidateAttributeSchema(record.schema) || record.recordId.empty() || record.trackId.empty() ||
        record.schemaId != AttributeSchemaId(record.schema) || record.firstSeen < 0 ||
        record.lastSeen < record.firstSeen || record.attributes.size() != record.schema.attributes.size())
        return false;
    std::set<std::string> keys;
    bool complete = true;
    for (const auto& value : record.attributes) {
        const auto definition = std::find_if(record.schema.attributes.begin(), record.schema.attributes.end(),
                                             [&](const auto& item) { return item.key == value.key; });
        if (definition == record.schema.attributes.end() || !keys.insert(value.key).second ||
            value.samples < 0 || !std::isfinite(value.confidence) || value.confidence < 0 ||
            value.confidence > 1)
            return false;
        if (value.status != "valid") {
            complete = false;
            if ((value.status != "unknown" && value.status != "failed") || !value.values.empty())
                return false;
            continue;
        }
        if (value.samples == 0 || value.values.empty() ||
            (definition->type != "multiple" && value.values.size() != 1))
            return false;
        std::set<std::string> values;
        for (const auto& option : value.values) {
            if (!values.insert(option).second ||
                std::none_of(definition->options.begin(), definition->options.end(),
                             [&](const auto& item) { return item.value == option; }))
                return false;
        }
    }
    return record.status == (complete ? "complete" : "partial");
}

bool ValidateAttributeFlow(const std::string& process, bool requireAttributes) {
    if (process.empty())
        return !requireAttributes;
    try {
        const auto nodes = nlohmann::json::parse(process);
        if (!nodes.is_array())
            return false;
        auto param = [](const nlohmann::json& node, const std::string& key) -> std::string {
            if (!node.contains("configObject"))
                return {};
            for (const auto& value : node.at("configObject").value("params", nlohmann::json::array())) {
                if (value.value("key", "") == key)
                    return value.value("value", "");
            }
            return {};
        };
        std::map<std::string, const nlohmann::json*> byId;
        const nlohmann::json* accumulation = nullptr;
        for (const auto& node : nodes) {
            if (node.value("actionId", "") == "BA_20003" && param(node, "inputMode") == "attributes") {
                if (accumulation)
                    return false;
                accumulation = &node;
            }
        }
        if (!accumulation)
            return !requireAttributes;
        for (const auto& node : nodes) {
            if (!byId.emplace(node.at("flowActionId").get<std::string>(), &node).second)
                return false;
        }
        const auto schema =
            nlohmann::json::parse(param(*accumulation, "attributeSchema")).get<AttributeSchema>();
        if (!ValidateAttributeSchema(schema))
            return false;
        auto ancestors = [&](const nlohmann::json& node) {
            std::set<std::string> seen;
            std::function<void(const nlohmann::json&)> visit = [&](const auto& current) {
                std::istringstream parents(current.value("preFlowActionId", ""));
                std::string id;
                while (std::getline(parents, id, ',')) {
                    if (byId.count(id) && seen.insert(id).second)
                        visit(*byId.at(id));
                }
            };
            visit(node);
            return seen;
        };
        const auto upstream = ancestors(*accumulation);
        // The first release accepts one sequential observation path. Fan-in alone
        // does not merge per-target classifier evidence from independent branches.
        for (const auto& id : upstream) {
            if (byId.at(id)->value("preFlowActionId", "").find(',') != std::string::npos)
                return false;
        }
        if (accumulation->value("preFlowActionId", "").find(',') != std::string::npos)
            return false;
        for (const auto& definition : schema.attributes) {
            if (!upstream.count(definition.sourceNode))
                return false;
            const auto& source   = *byId.at(definition.sourceNode);
            const auto preceding = ancestors(source);
            if (std::none_of(preceding.begin(), preceding.end(), [&](const auto& id) {
                    return byId.at(id)->value("actionId", "") == "AA_00003";
                }))
                return false;
            if (source.value("actionId", "") != "AA_00002" ||
                param(source, "atomicCode") != definition.modelCode)
                return false;
        }
        for (const auto& node : nodes) {
            if (node.value("actionId", "") == "BA_00004" &&
                ancestors(node).count(accumulation->at("flowActionId").get<std::string>()))
                return true;
        }
        return false;
    } catch (const nlohmann::json::exception&) {
        return false;
    }
}

std::string AttributeFingerprint(const std::string& value) {
    // Stable content fingerprint (not a security signature).
    uint64_t hash = 14695981039346656037ULL;
    for (unsigned char c : value) {
        hash ^= c;
        hash *= 1099511628211ULL;
    }
    std::ostringstream out;
    out << std::hex << std::setw(16) << std::setfill('0') << hash;
    return out.str();
}
}  // namespace cosmo
